File: 1485-minimum-cost-to-make-at-least-one-valid-path-in-a-grid/minimum-cost-to-make-at-least-one-valid-path-in-a-grid.java
class Solution {
    public int minCost(int[][] g) {
        int n = g.length ;
        int m = g[0].length ;

        int[][] di = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        int dist[][] = new int[n][m] ;
        for(int[] row : dist){
            Arrays.fill(row, Integer.MAX_VALUE) ;
        }
        dist[0][0] = 0 ;

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        pq.offer(new int[]{0, 0, 0});

        while(!pq.isEmpty()){
            int[] cr = pq.poll() ;
            int cost = cr[0];
            int x = cr[1] ;
            int y = cr[2] ;

            if(x == n - 1 && y == m - 1){
                return cost ;
            }

            for(int i = 0 ; i < 4 ; i ++){
                int nx = x + di[i][0] ;
                int ny = y + di[i][1] ;
                
                if(nx >= 0 && nx < n && ny >= 0 && ny < m){

                    int nwCost = cost + (i + 1 == g[x][y] ? 0 : 1);
                    if(nwCost < dist[nx][ny]){
                        dist[nx][ny] = nwCost ;
                        pq.offer(new int[]{nwCost, nx, ny}) ;
                    }
                }
            }
        }

        return -1 ;
    }
}

File: 1485-minimum-cost-to-make-at-least-one-valid-path-in-a-grid/minimum-cost-to-make-at-least-one-valid-path-in-a-grid.py
import heapq
class Solution:
    def minCost(self, g: List[List[int]]) -> int:
        q = []
        n, m = len(g), len(g[0])
        dist = [[float('inf') for _ in range(m)] for _ in range(n)]
        dist[0][0] = 0 
        heapq.heappush(q, (0, 0, 0))  # Push (cost, row, col) into the heap

        # Directions corresponding to the grid values
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        
        while q:
            cost, x, y = heapq.heappop(q)
            if x == n - 1 and y == m - 1:
                return cost 
            
            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if 0 <= nx < n and 0 <= ny < m:
                    new_cost = cost + (1 if (dx, dy) != (directions[g[x][y] - 1]) else 0)
                    if new_cost < dist[nx][ny]: 
                        dist[nx][ny] = new_cost
                        heapq.heappush(q, (new_cost, nx, ny))
        
        return -1 